Support for GHC-8.8

As far as I can tell, there are two things blocking `hedgehog`'s compatibility with GHC-8.8.1:

* The bound on `template-haskell` needs to be lifted

* Happy needs updates to its `Setup.hs`, which it appears to be getting in https://github.com/simonmar/happy/pull/145.

I just ran into this issue when trying to build Hedgehog with ``--allow-newer`` for another project of mine. I'm not sure this is covered by the fix(es) above.

```
Failed to build hedgehog-1.0.
Build log (
/home/koz/.cabal/logs/ghc-8.8.1/hedgehog-1.0-6b7004c5d78246ab13147530802b1ee0ff504798be00b7ac10195f734a78644a.log
):
Configuring library for hedgehog-1.0..
Preprocessing library for hedgehog-1.0..
Building library for hedgehog-1.0..
[ 1 of 25] Compiling Hedgehog.Internal.Config ( src/Hedgehog/Internal/Config.hs, dist/build/Hedgehog/Internal/Config.o )
[ 2 of 25] Compiling Hedgehog.Internal.Distributive ( src/Hedgehog/Internal/Distributive.hs, dist/build/Hedgehog/Internal/Distributive.o )
[ 3 of 25] Compiling Hedgehog.Internal.Exception ( src/Hedgehog/Internal/Exception.hs, dist/build/Hedgehog/Internal/Exception.o )
[ 4 of 25] Compiling Hedgehog.Internal.HTraversable ( src/Hedgehog/Internal/HTraversable.hs, dist/build/Hedgehog/Internal/HTraversable.o )
[ 5 of 25] Compiling Hedgehog.Internal.Opaque ( src/Hedgehog/Internal/Opaque.hs, dist/build/Hedgehog/Internal/Opaque.o )
[ 6 of 25] Compiling Hedgehog.Internal.Queue ( src/Hedgehog/Internal/Queue.hs, dist/build/Hedgehog/Internal/Queue.o )
[ 7 of 25] Compiling Hedgehog.Internal.Range ( src/Hedgehog/Internal/Range.hs, dist/build/Hedgehog/Internal/Range.o )
[ 8 of 25] Compiling Hedgehog.Internal.Region ( src/Hedgehog/Internal/Region.hs, dist/build/Hedgehog/Internal/Region.o )
[ 9 of 25] Compiling Hedgehog.Internal.Seed ( src/Hedgehog/Internal/Seed.hs, dist/build/Hedgehog/Internal/Seed.o )
[10 of 25] Compiling Hedgehog.Internal.Show ( src/Hedgehog/Internal/Show.hs, dist/build/Hedgehog/Internal/Show.o )
[11 of 25] Compiling Hedgehog.Internal.Shrink ( src/Hedgehog/Internal/Shrink.hs, dist/build/Hedgehog/Internal/Shrink.o )
[12 of 25] Compiling Hedgehog.Internal.Source ( src/Hedgehog/Internal/Source.hs, dist/build/Hedgehog/Internal/Source.o )
[13 of 25] Compiling Hedgehog.Internal.Tree ( src/Hedgehog/Internal/Tree.hs, dist/build/Hedgehog/Internal/Tree.o )
[14 of 25] Compiling Hedgehog.Main    ( src/Hedgehog/Main.hs, dist/build/Hedgehog/Main.o )
[15 of 25] Compiling Hedgehog.Range   ( src/Hedgehog/Range.hs, dist/build/Hedgehog/Range.o )
[16 of 25] Compiling Hedgehog.Internal.Gen ( src/Hedgehog/Internal/Gen.hs, dist/build/Hedgehog/Internal/Gen.o )

src/Hedgehog/Internal/Gen.hs:489:3: error:
    ‘fail’ is not a (visible) method of class ‘Monad’
    |
489 |   fail =
    |   ^^^^
```

--- a/.appveyor.yml
+++ b/.appveyor.yml
@@ -15,7 +15,8 @@ environment:
     - GHC_VERSION: "8.0.2"
     - GHC_VERSION: "8.2.2"
     - GHC_VERSION: "8.4.4"
-    - GHC_VERSION: "8.6.4"
+    - GHC_VERSION: "8.6.5"
+    - GHC_VERSION: "8.8.1"
 
 cache:
  - "C:\\SR"

--- a/.travis.yml
+++ b/.travis.yml
@@ -37,8 +37,10 @@ matrix:
       addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.4,ghc-8.2.2], sources: [hvr-ghc]}}
     - compiler: "ghc-8.4.4"
       addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.4,ghc-8.4.4], sources: [hvr-ghc]}}
-    - compiler: "ghc-8.6.4"
-      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.4,ghc-8.6.4], sources: [hvr-ghc]}}
+    - compiler: "ghc-8.6.5"
+      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.4,ghc-8.6.5], sources: [hvr-ghc]}}
+    - compiler: "ghc-8.8.1"
+      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-3.0,ghc-8.8.1], sources: [hvr-ghc]}}
 
 before_install:
   - HC=/opt/ghc/bin/${CC}

--- a/hedgehog/hedgehog.cabal
+++ b/hedgehog/hedgehog.cabal
@@ -67,7 +67,7 @@ library
     , resourcet                       >= 1.1        && < 1.3
     , semigroups                      >= 0.16       && < 0.20
     , stm                             >= 2.4        && < 2.6
-    , template-haskell                >= 2.10       && < 2.15
+    , template-haskell                >= 2.10       && < 2.16
     , text                            >= 1.1        && < 1.3
     , time                            >= 1.4        && < 1.10
     , transformers                    >= 0.5        && < 0.6

--- a/hedgehog/src/Hedgehog/Internal/Gen.hs
+++ b/hedgehog/src/Hedgehog/Internal/Gen.hs
@@ -1,5 +1,6 @@
 {-# OPTIONS_HADDOCK not-home #-}
 {-# LANGUAGE ApplicativeDo #-}
+{-# LANGUAGE CPP #-}
 {-# LANGUAGE DataKinds #-}
 {-# LANGUAGE DeriveFoldable #-}
 {-# LANGUAGE DeriveFunctor #-}
@@ -486,8 +487,11 @@ instance Monad m => Monad (GenT m) where
           runGenT size sk . k =<<
           runGenT size sm m
 
+#if MIN_VERSION_base(4,13,0)
+#else
   fail =
     Fail.fail
+#endif
 
 instance Monad m => MonadFail (GenT m) where
   fail =

